pkg/restic/backupper.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2018 the Heptio Ark contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restic
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	corev1api "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+
+	arkv1api "github.com/heptio/ark/pkg/apis/ark/v1"
+	"github.com/heptio/ark/pkg/util/boolptr"
+)
+
+// Backupper can execute restic backups of volumes in a pod.
+type Backupper interface {
+	// BackupPodVolumes backs up all annotated volumes in a pod.
+	BackupPodVolumes(backup *arkv1api.Backup, pod *corev1api.Pod, log logrus.FieldLogger) (map[string]string, []error)
+}
+
+type backupper struct {
+	repoManager *repositoryManager
+	informer    cache.SharedIndexInformer
+	results     map[string]chan *arkv1api.PodVolumeBackup
+	resultsLock sync.Mutex
+	ctx         context.Context
+}
+
+func resultsKey(ns, name string) string {
+	return fmt.Sprintf("%s/%s", ns, name)
+}
+
+func (b *backupper) BackupPodVolumes(backup *arkv1api.Backup, pod *corev1api.Pod, log logrus.FieldLogger) (map[string]string, []error) {
+	// get volumes to backup from pod's annotations
+	volumesToBackup := GetVolumesToBackup(pod)
+	if len(volumesToBackup) == 0 {
+		return nil, nil
+	}
+
+	// ensure a repo exists for the pod's namespace
+	if err := b.repoManager.ensureRepo(pod.Namespace); err != nil {
+		return nil, []error{err}
+	}
+
+	b.resultsLock.Lock()
+	b.results[resultsKey(pod.Namespace, pod.Name)] = make(chan *arkv1api.PodVolumeBackup)
+	b.resultsLock.Unlock()
+
+	var (
+		errs            []error
+		volumeSnapshots = make(map[string]string)
+	)
+
+	for _, volumeName := range volumesToBackup {
+		b.repoManager.repoLocker.Lock(pod.Namespace, false)
```

Comment:
how come?

Reply:
If we lock here, and we're backing up a namespace that has 20, 50, 100+ pods, we are repeatedly locking and unlocking it. If repo maintenance is about to trigger, I'd rather wait until the entire backup is done instead of interleaving it into backing up pods. WDYT?